chainer/links/connection/convolution_2d.py

```diff
@@ -167,6 +167,53 @@ def _initialize_params(self, in_channels):
         W_shape = (self.out_channels, int(in_channels / self.groups), kh, kw)
         self.W.initialize(W_shape)
 
+    @classmethod
+    def from_params(cls, W, b=None, stride=1, pad=0, nobias=False, **kwargs):
+        """from_params(cls, W, b=None, stride=1, pad=0, nobias=False, \
+*, dilate=1, groups=1)
+
+        Initialize a :class:`~chainer.links.Convolution2D` with given
+        parameters.
+
+        This method uses ``W`` and optional ``b`` to initialize
+        a 2D convolution layer.
+
+        Args:
+            W (:class:`~chainer.Variable` or :ref:`ndarray`):
+                The weight parameter.
+            b (:class:`~chainer.Variable`, :ref:`ndarray`, or ``None``):
+                The bias parameter.
+            stride (int or pair of ints): Stride of filter applications.
+                ``stride=s`` and ``stride=(s, s)`` are equivalent.
+            pad (int or pair of ints): Spatial padding width for input arrays.
+                ``pad=p`` and ``pad=(p, p)`` are equivalent.
+            nobias (bool): If ``True``, then this link does not use
+                the bias term in spite of whether ``b`` is given or not.
+            dilate (int or pair of ints):
+                Dilation factor of filter applications.
+                ``dilate=d`` and ``dilate=(d, d)`` are equivalent.
+            groups (:class:`int`): Number of groups of channels. If the number
+                is greater than 1, input tensor :math:`W` is divided into some
+                blocks by this value channel-wise. For each tensor blocks,
+                convolution operation will be executed independently.
+                Input channel size ``in_channels`` and output channel size
+                ``out_channels`` must be exactly divisible by this value.
+        """
+        dilate, groups = argument.parse_kwargs(
+            kwargs, ('dilate', 1), ('groups', 1))
+        out_channels, _in_channels, kw, kh = W.shape
+        in_channels = _in_channels * groups
+        if b is not None:
+            if out_channels != b.size:
+                raise ValueError(
+                    '`out_channels` does not match the size of `b`')
+
+        link = cls(
+            in_channels, out_channels, (kw, kh), stride, pad, nobias,
+            initialW=variable.as_array(W), initial_bias=variable.as_array(b),
```

Comment:
this code fails when arrays other than numpy are passed as initializers.

Running your branch in my environment shows 
```
 Base test method: TestConvolution2D_use_chainerx_false__chainerx_device_None__use_cuda_true__cuda_device_0__use_cudnn_always__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never.test_from_params
E           Test parameters:
E             W_dtype: <class 'numpy.float64'>
E             x_dtype: <class 'numpy.float64'>
E
E
E           (caused by)
E           TypeError: invalid type of initializer: <class 'cupy.core.core.ndarray'>
```

Reply:
I just submitted a PR to tackle this issue #8022